ledger/eval/eval.go

```diff
@@ -1380,24 +1415,39 @@ func (eval *BlockEvaluator) generateExpiredOnlineAccountsList() {
 	currentRound := eval.Round()
 
 	expectedMaxNumberOfExpiredAccounts := eval.proto.MaxProposedExpiredOnlineAccounts
+	expectedMaxNumberOfAbsentAccounts := eval.proto.MaxProposedAbsentOnlineAccounts
+
+	updates := &eval.block.ParticipationUpdates
 
-	for i := 0; i < len(modifiedAccounts) && len(eval.block.ParticipationUpdates.ExpiredParticipationAccounts) < expectedMaxNumberOfExpiredAccounts; i++ {
+	for i := 0; i < len(modifiedAccounts); i++ {
 		accountAddr := modifiedAccounts[i]
 		acctDelta, found := eval.state.mods.Accts.GetData(accountAddr)
 		if !found {
 			continue
 		}
 
-		// true if the account is online
-		isOnline := acctDelta.Status == basics.Online
-		// true if the accounts last valid round has passed
-		pastCurrentRound := acctDelta.VoteLastValid < currentRound
-
-		if isOnline && pastCurrentRound {
-			eval.block.ParticipationUpdates.ExpiredParticipationAccounts = append(
-				eval.block.ParticipationUpdates.ExpiredParticipationAccounts,
-				accountAddr,
-			)
+		if acctDelta.Status == basics.Online {
+			expiresBeforeCurrent := acctDelta.VoteLastValid < currentRound
+			if expiresBeforeCurrent &&
+				len(updates.ExpiredParticipationAccounts) < expectedMaxNumberOfExpiredAccounts {
+				updates.ExpiredParticipationAccounts = append(
+					updates.ExpiredParticipationAccounts,
+					accountAddr,
+				)
+			}
+			// TODO: 1000 rounds is obviously a placeholder. Probably needs to
+			// be something like 10x the "expected" interval.  Further
+			// complications: (1) At keyreg, LastProposed=0 (or is old). (2)
+			// Accounts should be allowed to "heartbeat" to stay online. (3)
+			// "Expected" interval varies with balance and onlinestake.
+			lastVoteTooOld := acctDelta.LastProposed+1000 < currentRound
+			if lastVoteTooOld &&
+				len(updates.AbsentParticipationAccounts) < expectedMaxNumberOfAbsentAccounts {
+				updates.AbsentParticipationAccounts = append(
+					updates.AbsentParticipationAccounts,
+					accountAddr,
+				)
+			}
```

Comment:
so an account can be both expired and absent

Reply:
Right, "expired" is a function of the parameters you set when you `keyreg`d  (and is _not set as a `Status`, you are still `Online`) and "absent" is a function of your performance (and _is_ a new `Status`).  If you expire, but don't explicitly go offline (and you don't get knocked off), you will certainly also end up "absent", since you can't participate.

As written here, both of these detections would only occur when a transaction that involves your account hits the chain. So good samaritans would note your absence or expiration and send you 0 algos, so the next proposer would notice you. Alternatively, `algod` _could_ try harder to find absent or expired accounts. We just don't want it doing a full account scan to find them.

There are probably some details to be worked out to make sure that, for example, a node wouldn't mark the same account expired and suspended in the same block, or, ensure that recipients of such a block would not fail. (I'm going to do both, now)

Reply:
FWIW, I'm rethinking the creation of a new Status (`Suspended`). I think it may be a smaller, safer change to imply suspension with Offline status, but without zero'd Voting Keys.  As I've made updates to handle a suspended flag, it seems like each case would have been handled naturally with that change instead.

Reply:
`Suspended` is now gone as an explicit status. Instead, it's implicit - when an acocunt is suspended, its status become `Offline`, but it's key material is retained, so that a heartbeat or proposal can bring it back online.

It may be surprising that I said in that last sentence. How can it propose if it's `Offline`? Like all changes to `Offline` status, the account is still eligible to vote for 320 rounds. So an account could be suspended and then come back `Online` by proposing. But it should probably heartbeat instead, so that it is not `Offline` for a small gap as it's suspension becomes 320 rounds old.